parser.y

```diff
@@ -2312,74 +2411,131 @@ PartitionDefinitionList:
 	}
 
 PartitionDefinition:
-	"PARTITION" Identifier PartDefValuesOpt PartDefOptionsOpt
+	"PARTITION" Identifier PartDefValuesOpt PartDefOptionList SubPartDefinitionListOpt
 	{
-		partDef := &ast.PartitionDefinition{
-			Name: model.NewCIStr($2),
-		}
-		switch $3.(type) {
-		case []ast.ExprNode:
-			partDef.LessThan = $3.([]ast.ExprNode)
-		case ast.ExprNode:
-			partDef.LessThan = make([]ast.ExprNode, 1)
-			partDef.LessThan[0] = $3.(ast.ExprNode)
+		$$ = &ast.PartitionDefinition{
+			Name:    model.NewCIStr($2),
+			Clause:  $3.(ast.PartitionDefinitionClause),
+			Options: $4.([]*ast.TableOption),
+			Sub:     $5.([]*ast.SubPartitionDefinition),
 		}
+	}
 
-		if comment, ok := $4.(string); ok {
-			partDef.Comment = comment
-		}
-		$$ = partDef
+SubPartDefinitionListOpt:
+	/*empty*/
+	{
+		$$ = make([]*ast.SubPartitionDefinition, 0)
+	}
+|	'(' SubPartDefinitionList ')'
+	{
+		$$ = $2
 	}
 
-PartDefOptionsOpt:
+SubPartDefinitionList:
+	SubPartDefinition
 	{
-		$$ = nil
+		$$ = []*ast.SubPartitionDefinition{$1.(*ast.SubPartitionDefinition)}
 	}
-|	PartDefOptionList
+|	SubPartDefinitionList ',' SubPartDefinition
 	{
-		$$ = $1
+		list := $1.([]*ast.SubPartitionDefinition)
+		$$ = append(list, $3.(*ast.SubPartitionDefinition))
+	}
+
+SubPartDefinition:
+	"SUBPARTITION" Identifier PartDefOptionList
+	{
+		$$ = &ast.SubPartitionDefinition{
+			Name:    model.NewCIStr($2),
+			Options: $3.([]*ast.TableOption),
+		}
 	}
 
 PartDefOptionList:
-	PartDefOption
+	/*empty*/
 	{
-		$$ = $1
+		$$ = make([]*ast.TableOption, 0)
 	}
 |	PartDefOptionList PartDefOption
 	{
-		if $1 != nil {
-			$$ = $1
-		} else {
-			$$ = $2
-		}
+		list := $1.([]*ast.TableOption)
+		$$ = append(list, $2.(*ast.TableOption))
 	}
 
 PartDefOption:
 	"COMMENT" EqOpt stringLit
 	{
-		$$ = $3
+		$$ = &ast.TableOption{Tp: ast.TableOptionComment, StrValue: $3}
 	}
-|	"ENGINE" EqOpt Identifier
+|	"ENGINE" EqOpt StringName
```

Comment:
Is there a particular reason why changing `Identifier` to `StringName`?

Reply:
@bb7133 It is valid to use a string here. Also make it possible to share the same rule for both table options and partition options.

```sql
create table t(a int) partition by hash(a) (partition `x` engine = 'innodb');
```

Reply:
Thank you!